Replaced the DECLARE_VIRTUAL_REFCOUNTED macro with an abstract base class

geoffreygaren opened the pull request:
#### 59ca354ec983a0f06f75fb8bdb0983d88eb1c785
<pre>
Replaced the DECLARE_VIRTUAL_REFCOUNTED macro with an abstract base class
<a href="https://bugs.webkit.org/show_bug.cgi?id=281428">https://bugs.webkit.org/show_bug.cgi?id=281428</a>
<a href="https://rdar.apple.com/137875169">rdar://137875169</a>

Reviewed by Chris Dumez.

Chris pointed out that macros are hard to read/understand, and can interfere
with debugging.

We can use an abstract base class instead.

I like having a class for this because it matches how we declare concrete
refcounting.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FunctionDispatcher.h:
* Source/WTF/wtf/RefCounted.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.h:
* Source/WebCore/Modules/push-api/PushSubscriptionOwner.h:
* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSRuleList.h:
* Source/WebCore/css/CSSStyleDeclaration.h:
* Source/WebCore/dom/ActiveDOMObject.h:
* Source/WebCore/loader/ContentFilterClient.h:
* Source/WebCore/page/SpeechSynthesisClient.h:
* Source/WebCore/platform/DateTimeChooser.h:
* Source/WebCore/platform/RemoteCommandListener.h:
* Source/WebCore/platform/audio/AudioDestination.h:
* Source/WebCore/platform/audio/AudioHardwareListener.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:
(WebCore::CapturerObserver::capturerConfigurationChanged):
* Source/WebCore/platform/network/curl/CurlRequestClient.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordLoader.h:
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.h:
* Source/WebKit/UIProcess/FrameLoadState.h:
* Source/WebKit/UIProcess/PageLoadState.h:
* Source/WebKit/UIProcess/ResponsivenessTimer.h:
* Source/WebKit/UIProcess/WebAuthentication/Authenticator.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
* Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.h:
* Source/WebKit/webpushd/PushService.mm:

Canonical link: <a href="https://commits.webkit.org/285138@main">https://commits.webkit.org/285138@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9261aeae7d39a407d230fd3acff3f3571842154

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51068 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75769 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22860 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73770 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58869 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22680 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56596 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15079 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74721 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46333 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61726 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37045 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19193 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21201 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64780 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64898 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19556 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77489 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70905 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15889 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18736 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15932 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61759 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64307 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12458 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6098 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92690 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10987 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46868 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1647 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20434 "Failed to compile JSC") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47939 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49223 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47681 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/FunctionDispatcher.h:
```diff
@@ -52,7 +52,8 @@ class WTF_CAPABILITY("is current") WTF_EXPORT_PRIVATE SerialFunctionDispatcher :
 class RefCountedSerialFunctionDispatcher : public SerialFunctionDispatcher {
 public:
-    DECLARE_VIRTUAL_REFCOUNTED;
+    virtual void ref() const = 0;
+    virtual void deref() const = 0;
```
I skipped the base class here for now because I got a cryptic linker error about

> no declaration found for exported symbol 'non-virtual thunk...'

that I couldn't figure out.

cdumez commented on Source/WTF/wtf/AbstractRefCounted.h:
```diff
@@ -0,0 +1,41 @@
+
+namespace WTF {
+
+// Use this class when an asbstract base class needs refcounting, and the
```
typo: `asbstract`

cdumez commented on Source/WTF/wtf/AbstractRefCounted.h:
```diff
@@ -0,0 +1,41 @@
+// refcounting implementation will be in a concrete subclass.
+class AbstractRefCounted {
+public:
+    virtual ~AbstractRefCounted() = default;
```
Should this destructor be `protected`?

cdumez commented on Source/WebCore/css/CSSStyleDeclaration.h:
```diff
@@ -36,12 +37,10 @@ class MutableStyleProperties;
 class StyledElement;
 
-class CSSStyleDeclaration : public ScriptWrappable, public CanMakeSingleThreadWeakPtr<CSSStyleDeclaration> {
+class CSSStyleDeclaration : public AbstractRefCountedAndCanMakeSingleThreadWeakPtr<CSSStyleDeclaration>, public ScriptWrappable {
```
A little worried about swapping the order of base classes here. Having `ScriptWrappable` as first base class made more sense to me (though hopefully the bindings code doesn't rely on that?)

geoffreygaren committed:
Replaced the DECLARE_VIRTUAL_REFCOUNTED macro with an abstract base class
https://bugs.webkit.org/show_bug.cgi?id=281428
rdar://137875169

Reviewed by Chris Dumez.

Chris pointed out that macros are hard to read/understand, and can interfere
with debugging.

We can use an abstract base class instead.

I like having a class for this because it matches how we declare concrete
refcounting.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FunctionDispatcher.h:
* Source/WTF/wtf/RefCounted.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.h:
* Source/WebCore/Modules/push-api/PushSubscriptionOwner.h:
* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSRuleList.h:
* Source/WebCore/css/CSSStyleDeclaration.h:
* Source/WebCore/dom/ActiveDOMObject.h:
* Source/WebCore/loader/ContentFilterClient.h:
* Source/WebCore/page/SpeechSynthesisClient.h:
* Source/WebCore/platform/DateTimeChooser.h:
* Source/WebCore/platform/RemoteCommandListener.h:
* Source/WebCore/platform/audio/AudioDestination.h:
* Source/WebCore/platform/audio/AudioHardwareListener.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
* Source/WebCore/platform/graphics/avfoundation/SampleBufferDisplayLayer.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:
(WebCore::CapturerObserver::capturerConfigurationChanged):
* Source/WebCore/platform/network/curl/CurlRequestClient.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordLoader.h:
* Source/WebCore/workers/service/context/SWContextManager.h:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/UIProcess/Cocoa/ExtensionCapabilityGranter.h:
* Source/WebKit/UIProcess/FrameLoadState.h:
* Source/WebKit/UIProcess/PageLoadState.h:
* Source/WebKit/UIProcess/ResponsivenessTimer.h:
* Source/WebKit/UIProcess/WebAuthentication/Authenticator.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
* Source/WebKit/WebProcess/Network/webrtc/WebRTCMonitor.h:
* Source/WebKit/webpushd/PushService.mm:

Canonical link: https://commits.webkit.org/285138@main
